### PR TITLE
fix(continuous-batching): apply logits processors in packed batches

### PR DESCRIPTION
**Pull request by floor-licker**
## Summary
It seems like this was a known issue, but the continuous batching implementation packs multiple sequences into a single token stream even though most generation-time logits processors assume `[batch, seq_len]` / `[batch, vocab]` shapes. In practice it looks like processors are either broken or disabled in `transformers serve` and the tests. This change would apply logits processors per-request on the correct next token logits positions and removes the workarounds. 

Logic: 
  - Apply logits processors only at “next token” positions indicated by
  `logits_indices`, and only for requests that are in decode
  - For each decoding request, rebuild the per-request token history
  (`state.initial_tokens + state.generated_tokens`) and apply the processor to
  `logits[0, position]` (next-token scores) in-place

  ## Tests run
  - python -m pytest tests/generation/test_continuous_batching.py -q
  - python -m pytest tests/generation/test_logits_process.py -k
  repetition_penalty_continuous_batching -q
  - python -m pytest tests/cli/test_serve.py -q
  - RUN_SLOW=1 python -m pytest tests/cli/test_serve.py -k
  ServeCompletionsContinuousBatchingIntegrationTest -q

### CODE REVIEW

**Comment by jtranq** [comment]
Some of these CI failures seem a bit odd relative to the code changes, may take me a minute to track this down.

**Comment by jtranq** [comment]
Hey @LysandreJik , just wondering if I can get some feedback on this. As far as I can tell the failing tests are just flaky. Can only intermittently reproduce the failures locally.

**Comment by Rocketknight1** [comment]
cc @mcpatate @remi-or for continuous batching

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 1]
```diff
@@ -761,19 +761,44 @@ def _model_forward(self, model: nn.Module, batch_data: dict) -> torch.Tensor:
 
     @traced(span_name="logit_processing")
     def _process_logit(self, batch_data: dict, logits: torch.Tensor, logit_processor: LogitsProcessor) -> torch.Tensor:
-        # Pass continuous batching context to logits processor if it supports it.
+        if isinstance(logit_processor, list) and len(logit_processor) == 0:
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/continuous_api.py:764]: This should always be a list: you can check `_get_logits_processor`

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 2]
```diff
@@ -761,19 +761,44 @@ def _model_forward(self, model: nn.Module, batch_data: dict) -> torch.Tensor:
...
+        logits_indices = batch_data["logits_indices"]
+        # -1 means prefill.
+        next_token_positions = logits_indices[logits_indices >= 0]
+        if next_token_positions.numel() == 0:
+            # Nothing to process.
+            return logits
+
+        # Decode-only states.
+        states_for_next_token = [state for state in self.requests_in_batch if len(state.remaining_prefill_tokens) == 0]
+        if len(states_for_next_token) != next_token_positions.numel():
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/continuous_api.py]: Has this been a problem in your experience? It should not happen imo

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 3]
```diff
@@ -761,19 +761,44 @@ def _model_forward(self, model: nn.Module, batch_data: dict) -> torch.Tensor:
...
+            next_token_positions = next_token_positions[:length]
+
+        for state, position in zip(states_for_next_token, next_token_positions):
+            position = position.item()
+            # Rebuild token history.
+            input_ids = torch.tensor(
+                state.initial_tokens + state.generated_tokens, device=logits.device, dtype=torch.long
+            ).unsqueeze(0)
+            # Slice next-token logits.
+            scores = logits[0, position].unsqueeze(0)
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/continuous_api.py]: No need to do [0, ...] then `unsqueeze(0)`

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 4]
```diff
@@ -853,6 +878,20 @@ def __init__(
             num_kv_padding_intervals=num_kv_padding_intervals,
             compile_config=getattr(generation_config, "compile_config", None),
         )
+        if isinstance(self.logit_processor, list) and len(self.logit_processor) > 0:
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/continuous_api.py]: cf. above

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 5]
```diff
@@ -853,6 +878,20 @@ def __init__(
...
+            # Processors need eager.
+            if self.use_cuda_graph:
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/continuous_api.py:878]: IMO this is not the optimal solution: the process_logits phase and what happens afterwards should be outside of the cuda graph / compile if there are processors. Do you think you could rewrite it that way?

File: tests/cli/test_serve.py [hunk 6]
```diff
@@ -656,12 +656,10 @@ def test_full_request(self):
...
-        # Verify that the system prompt went through.
-        self.assertTrue(
-            full_text.startswith(
-                "I can assist you with a wide range of tasks, from answering questions to providing information on various sports topics."
-            )
-        )
+        # System prompt applied.
+        # Expect "sports" mention.
+        self.assertTrue(full_text.strip())
+        self.assertIn("sports", full_text.lower())
```
Comment (remi-or) [inline tests/cli/test_serve.py:662]: Not sure why we should change this test!

File: tests/generation/test_continuous_batching.py [hunk 7]
```diff
@@ -137,6 +139,94 @@ def test_group_layers(
                     f"Test failed for: {layer_types_str = }, {sliding_window = }, {group_types = }",
                 )
 
+    def test_logits_processors_apply_to_next_token_logits(self) -> None:
+        vocab_size = 10
+        logits = torch.ones((1, 4, vocab_size), device=torch_device, dtype=torch.float)
+
+        # Req0 next-token @2.
+        logits[0, 2, 1] = -2.0
```
Comment (remi-or) [inline tests/generation/test_continuous_batching.py:147]: Seems like you can hardcode this whole part as one tensor declaration

File: tests/generation/test_continuous_batching.py [hunk 8]
```diff
@@ -137,6 +139,94 @@ def test_group_layers(
...
+        }
+
+        # Apply processing.
+        processed_logits = batch_processor._process_logit(batch_data, logits, logit_processor)
+
+        # Context propagated.
+        self.assertTrue(torch.equal(rep_penalty.logits_indices, batch_data["logits_indices"]))
+
+        # Req0 token penalties.
+        self.assertAlmostEqual(processed_logits[0, 2, 1].item(), -4.0)
```
Comment (remi-or) [inline tests/generation/test_continuous_batching.py:180]: Why not do bactched checks?

File: src/transformers/cli/serve.py [hunk 9]
```diff
@@ -1826,8 +1822,18 @@ def _load_model_and_data_processor(self, model_id_and_revision: str):
         architecture = getattr(transformers, config.architectures[0])
         model = architecture.from_pretrained(model_id, **model_kwargs)
 
+        # Default `max_length` is 20.
```
Comment (remi-or) [inline src/transformers/cli/serve.py:1825]: These comments are weird: did you use some AI assistant for this?

**Review by remi-or — COMMENTED** [review]
Thanks for the contribution! There are some large points to address here, as this is not a trivial part of the continuous batching code. It would be good if we can support this in the most optimized way possible. Let me know if my comments are clear!

**Comment by jtranq** [comment]
@remi-or Thanks a lot for your feedback, its my first contribution to transformers so I'll take a minute to review your comments and get back to you.

File: src/transformers/generation/continuous_batching/continuous_api.py (hunk 2 above)
Comment (jtranq) [reply src/transformers/generation/continuous_batching/continuous_api.py]: Was just defensive but you're right, there's an invariant there and `if len(states_for_next_token) != next_token_positions.numel()` is redundant.

File: src/transformers/generation/continuous_batching/continuous_api.py (hunk 3 above)
Comment (jtranq) [reply src/transformers/generation/continuous_batching/continuous_api.py]: agreed, thank you

**Comment by JLenzy** [comment]
Hello! I just discovered this PR as I'm looking to add per-request logit processors in a production continuous batching setting. @remi-or do you know if this is currently supported?

**Comment by remi-or** [comment]
> Hello! I just discovered this PR as I'm looking to add per-request logit processors in a production continuous batching setting. @remi-or do you know if this is currently supported?

I am not sure logits processors are supported in the first place, and I am sure per-request is not supported. Not sure it will ever be in production settings, where you need CUDA graphs, which makes per-request treatment hard in (most) cases. What logits processors do you have in mind? I can take a look.

**Comment by JLenzy** [comment]
> > Hello! I just discovered this PR as I'm looking to add per-request logit processors in a production continuous batching setting. @remi-or do you know if this is currently supported?
> 
> I am not sure logits processors are supported in the first place, and I am sure per-request is not supported. Not sure it will ever be in production settings, where you need CUDA graphs, which makes per-request treatment hard in (most) cases. What logits processors do you have in mind? I can take a look.

Thanks for your attention to this! 
My team is preparing to use continuous batching to deploy a model for widespread use. During internal testing, we have received a lot of requests for 'weirder' outputs - we work in generative music. In research environments (single batch generations), we have been able to expose Top-p/top-k/temperature inference settings, which has allowed for a lot of creative exploration. But my current understanding is that we need to instantiate a single GenerationConfig and then run that across the board for all generations?

**Comment by remi-or** [comment]
> But my current understanding is that we need to instantiate a single GenerationConfig and then run that across the board for all generations?

Yes, that's the case currently. Maybe we can complexify this for some logits processors, if pytorch operators permit it. That would only be those three parameters?

**Comment by JLenzy** [comment]
In an 'ideal world' we could modify per generation:
- top_p
- min_p
- top_k
- temperature
- repetition_penalty
- no_repeat_ngram_size

But top-p/k/temp as a starting point would already be immensely helpful. I would be happy to help contribute to this effort, bearing in mind that it'd be my first contribution to the HF ecosystem :)

**Comment by remi-or** [comment]
Hey @JLenzy , we are adding this feature in https://github.com/huggingface/transformers/pull/45026
The draft will probably be removed tomorrow. Dont hesitate to let us know if you have early feedback.

**Comment by remi-or** [comment]
Closed thanks to https://github.com/huggingface/transformers/pull/45026
